scripts/spm-integration-test.sh

```diff
@@ -0,0 +1,113 @@
+#!/bin/bash
+
+set -e -uxf -o pipefail
+wait_seconds=5
+retry_count=10
+compose_file=docker-compose/monitor/docker-compose.yml
+
+create_logs(){
+  echo "::group:: docker logs"
+  docker compose -f $compose_file logs
+  echo "::endgroup::"
+}
+
+# Function to check if a service is healthy
+check_service_health() {
+  local service_name=$1
+  local url=$2
+
+  echo "Checking health of service: $service_name at $url"
+  for i in $(seq 1 $retry_count); do
+    if curl -s -L --head --request GET "$url" | grep "200 OK" > /dev/null; then
+      echo "$service_name is healthy"
+      return 0
+    else
+      echo "Waiting for $service_name to be healthy... ($i/$retry_count)"
+      sleep $wait_seconds
+    fi
+  done
+
+  echo "Error: $service_name did not become healthy in time"
+  create_logs
+  return 1
+}
+
+# Function to check if all services are healthy
+wait_for_services() {
+  echo "Waiting for services to be up and running..."
+  check_service_health "Jaeger" "http://localhost:16686"
+  check_service_health "Prometheus" "http://localhost:9090/graph"
+  check_service_health "Grafana" "http://localhost:3000"
+}
+
+# Function to check SPM
+check_spm() {
+  local timeout=180
+  local interval=5
+  local end_time=$((SECONDS + timeout))
+  echo "Checking SPM"
+  services_list=("driver" "customer" "mysql" "redis" "frontend" "route" "ui")
+  for service in "${services_list[@]}"; do
+    echo "Processing service: $service"
+    while [ $SECONDS -lt $end_time ]; do
+      response=$(curl -s "http://localhost:16686/api/metrics/calls?service=$service&endTs=$(date +%s)000&lookback=1000&step=100&ratePer=60000")
+      service_name=$(echo "$response" | jq -r 'if .metrics and .metrics[0] then .metrics[0].labels[] | select(.name=="service_name") | .value else empty end')
+      if [ "$service_name" != "$service" ]; then
+        echo "Service name does not match '$service'"
+        sleep $interval
+      else
+        echo "Service name matched with '$service'"
+        break
+      fi
+    done
+    if [ $SECONDS -gt $end_time ]; then
+      echo "Error: $service service did not become healthy in time"
+      create_logs
```

Comment:
why not call this from teardown_services ?

Reply:
then it will also dump logs if the file is ran without any error do we still need to dump logs when there is no error ?

Reply:
it's not the end of the world, but if you want to suppress that there may be a way to check for exit code in trap function (if not you can always use a global variable in the script and only set it to succeed at the very end, to skip the logging).